hare/hare.go

```diff
@@ -0,0 +1,182 @@
+
+type consensusFactory func(cfg config.Config, key crypto.PublicKey, instanceId InstanceId, s *Set, oracle Rolacle, signing Signing, p2p NetworkService, terminationReport chan TerminationOutput) Consensus
+
+// Consensus represents a consensus
+type Consensus interface {
+	IdentifiableInboxer
+
+	Close()
+	CloseChannel() chan struct{}
+
+	Start() error
+}
+
+// TerminationOutput is a result of a process terminated with output.
+type TerminationOutput interface {
+	Id() []byte
+	Set() *Set
+}
+
+type orphanBlockProvider interface {
+	GetOrphanBlocks() []mesh.BlockID
+}
+
+// Hare is an orchestrator that shoots consensus processes and collects their termination output
+type Hare struct {
+	Closer
+
+	config config.Config
+
+	network    NetworkService
+	beginLayer chan mesh.LayerID
+
+	broker *Broker
+
+	pubKey crypto.PublicKey
+	sign   Signing
+
+	obp     orphanBlockProvider
+	rolacle Rolacle
+
+	networkDelta time.Duration
+
+	outputChan chan TerminationOutput
+	mu      sync.RWMutex
+	outputs map[mesh.LayerID][]mesh.BlockID
+
+	factory consensusFactory
+}
+
+// New returns a new Hare struct.
+func New(conf config.Config, p2p NetworkService, me crypto.PublicKey, sign Signing, obp orphanBlockProvider, rolacle Rolacle, beginLayer chan mesh.LayerID) *Hare {
+	h := new(Hare)
+	h.Closer = NewCloser()
+
+	h.config = conf
+
+	h.network = p2p
+	h.beginLayer = beginLayer
+
+	h.broker = NewBroker(p2p)
+
+	h.pubKey = me
+	h.sign = sign
+
+	h.obp = obp
+	h.rolacle = rolacle
+
+	h.networkDelta = Delta
+
+	h.outputChan = make(chan TerminationOutput, 20) // todo: figure out number of consensus processes can be run in parallel
+	h.outputs = make(map[mesh.LayerID][]mesh.BlockID)
+
+	h.factory = func(conf config.Config, key crypto.PublicKey, instanceId InstanceId, s *Set, oracle Rolacle, signing Signing, p2p NetworkService, terminationReport chan TerminationOutput) Consensus {
+		return NewConsensusProcess(conf, key, instanceId, s, oracle, signing, p2p, terminationReport)
+	}
+
+	return h
+}
+
+func (h *Hare) collectOutput(output TerminationOutput) {
+	id := output.Id()
+	set := output.Set()
+	blocks := make([]mesh.BlockID, len(set.values))
+	i := 0
+	for _, v := range set.values {
+		blocks[i] = mesh.BlockID(common.BytesToUint32(v.Bytes()))
+		i++
+	}
+	h.mu.Lock()
+	h.outputs[mesh.LayerID(common.BytesToUint32(id))] = blocks
+	h.mu.Unlock()
+}
+
+func (h *Hare) onTick(id mesh.LayerID) {
+	ti := time.NewTimer(h.networkDelta)
+	select {
+	case <-ti.C:
+		break // keep going
+	case <-h.CloseChannel():
+		// closed while waiting the delta
+		return
+	}
+
+	// retrieve set form orphan blocks
+	blocks := h.obp.GetOrphanBlocks()
+
+	set := NewEmptySet(len(blocks))
+
+	for _, b := range blocks {
+		// todo: figure out real type of blockid
+		set.Add(Value{NewBytes32(b.ToBytes())})
+	}
+
+	instid := InstanceId{NewBytes32(id.ToBytes())}
+
+	cp := h.factory(h.config, h.pubKey, instid, set, h.rolacle, h.sign, h.network, h.outputChan)
+	h.broker.Register(cp)
+	cp.Start()
```

Comment:
You should always start the consensus process before registering to the broker, or else you would be exposed to races (the broker will forward a message to a CP that is not listening yet).
Please do me a favor and add the following comment to the Register func in the broker:
"Note: the registering instance is assumed to be started and accepting messages"